src/coord/src/persistence.rs

```diff
@@ -444,3 +414,103 @@ fn augment_connector_inner(
 
     Ok(())
 }
+
+// Given the input records, extract a prefix of records that are "dense," meaning they contain all
+// the data for that range.
+fn extract_prefix(starting_offset: Option<i64>, records: &mut Vec<Record>) -> Vec<Record> {
+    records.sort_by(|a, b| (a.offset, a.predecessor.map(|p| -p)).cmp(&(b.offset, b.predecessor)));
+
+    // Keep only the minimum predecessor we received for every offset
+    records.dedup_by_key(|x| x.offset);
+    if let Some(offset) = starting_offset {
+        records.retain(|x| x.offset > offset);
+    }
+
+    let mut watermark = starting_offset;
+    let mut prefix_length = 0;
+
+    // Find the longest unbroken prefix where each subsequent record refers to the
+    // preceding one as its predecessor.
+    for p in records.iter() {
+        // Note: None < Some(x) for all x.
+        if p.predecessor <= watermark {
```

Comment:
I think I would feel better if we broke this out into
```
if p.predecessor.is_none() || p.predecessor == watermark
```

because if they are both none and p.predecessor is somehow less than watermark then I think something has gone very wrong and actually maybe we should throw some kind of error or something? does that sound right?

Reply:
wait actually I'm not sure about my suggestion either -- if the watermark is some(x) and the predecessor is none then not totally sure what we should do